File: android/src/main/java/com/pistadb/PistaDB.java
/*
 *    ___ _    _        ___  ___
 *   | _ (_)__| |_ __ _|   \| _ )
 *   |  _/ (_-<  _/ _` | |) | _ \
 *   |_| |_/__/\__\__,_|___/|___/
 */
package com.pistadb;

import java.io.Closeable;

/**
 * PistaDB – embedded vector database for Android.
 *
 * <p>Wraps the native C library through JNI.  Each instance holds an opaque
 * handle to a {@code PistaDB*} allocated on the native heap; call
 * {@link #close()} (or use try-with-resources) to free it.
 *
 * <h3>Quick-start (Java)</h3>
 * <pre>{@code
 * try (PistaDB db = new PistaDB(path, 384)) {
 *     db.insert(1L, embedding, "my document");
 *     SearchResult[] hits = db.search(queryVec, 5);
 *     db.save();
 * }
 * }</pre>
 *
 * <h3>Quick-start (Kotlin DSL)</h3>
 * <pre>{@code
 * val db = pistaDB(path, dim = 384) {
 *     metric    = Metric.COSINE
 *     indexType = IndexType.HNSW
 *     params.hnswEfSearch = 100
 * }
 * }</pre>
 *
 * <p><b>Thread safety:</b> all public methods are {@code synchronized}.
 * Concurrent access from multiple threads is safe but serialised.
 */
public class PistaDB implements Closeable {

    static {
        System.loadLibrary("pistadb_jni");
    }

    /* ── Native method declarations ──────────────────────────────────────── */

    private static native long   nativeOpen(String path, int dim,
                                            int metric, int indexType,
                                            PistaDBParams params);
    private static native void   nativeClose(long handle);
    private static native void   nativeSave(long handle);
    private static native void   nativeInsert(long handle, long id,
                                              String label, float[] vec);
    private static native void   nativeDelete(long handle, long id);
    private static native void   nativeUpdate(long handle, long id, float[] vec);
    private static native VectorEntry    nativeGet(long handle, long id);
    private static native SearchResult[] nativeSearch(long handle,
                                                      float[] query, int k);
    private static native void   nativeTrain(long handle);
    private static native int    nativeCount(long handle);
    private static native int    nativeDim(long handle);
    private static native int    nativeMetric(long handle);
    private static native int    nativeIndexType(long handle);
    private static native String nativeLastError(long handle);
    private static native String nativeVersion();

    /* ── State ───────────────────────────────────────────────────────────── */

    private volatile long nativeHandle;

    /* ── Construction ────────────────────────────────────────────────────── */

    /**
     * Opens or creates a database with default parameters and HNSW index.
     *
     * @param path File path for the {@code .pst} file.
     * @param dim  Vector dimensionality.
     */
    public PistaDB(String path, int dim) {
        this(path, dim, Metric.L2, IndexType.HNSW, null);
    }

    /**
     * Opens or creates a database.
     *
     * @param path      File path for the {@code .pst} file.
     * @param dim       Vector dimensionality.
     * @param metric    Distance metric.
     * @param indexType Index algorithm.
     * @param params    Tuning parameters; pass {@code null} for defaults.
     * @throws PistaDBException if the file cannot be opened or created.
     */
    public PistaDB(String path, int dim, Metric metric,
                   IndexType indexType, PistaDBParams params) {
        if (path == null)   throw new IllegalArgumentException("path must not be null");
        if (dim <= 0)       throw new IllegalArgumentException("dim must be > 0");
        if (metric == null) metric = Metric.L2;
        if (indexType == null) indexType = IndexType.HNSW;

        nativeHandle = nativeOpen(path, dim,
                                  metric.getValue(), indexType.getValue(),
                                  params != null ? params : PistaDBParams.defaults());
        if (nativeHandle == 0) {
            throw new PistaDBException("Failed to open database at: " + path);
        }
    }

    /* ── Lifecycle ───────────────────────────────────────────────────────── */

    /**
     * Saves the database to disk.  Does NOT close the handle.
     *
     * @throws PistaDBException on I/O error.
     */
    public synchronized void save() {
        checkOpen();
        nativeSave(nativeHandle);
    }

    /**
     * Saves the database, then frees the native handle.
     * Safe to call multiple times.
     */
    @Override
    public synchronized void close() {
        if (nativeHandle != 0) {
            try {
                nativeSave(nativeHandle);
            } catch (PistaDBException ignored) {
                // best-effort save; always free
            } finally {
                nativeClose(nativeHandle);
                nativeHandle = 0;
            }
        }
    }

    /* ── CRUD ────────────────────────────────────────────────────────────── */

    /**
     * Inserts a vector with an optional label.
     *
     * @param id    Unique identifier (user-managed).
     * @param vec   Float array of length {@link #getDim()}.
     * @param label Optional human-readable label (max 255 bytes); may be null.
     * @throws PistaDBException on duplicate id or dimension mismatch.
     */
    public synchronized void insert(long id, float[] vec, String label) {
        checkOpen();
        nativeInsert(nativeHandle, id, label != null ? label : "", vec);
    }

    /** Inserts a vector without a label. */
    public synchronized void insert(long id, float[] vec) {
        insert(id, vec, "");
    }

    /**
     * Soft-deletes a vector.  Space is reclaimed on the next {@link #save()}.
     *
     * @throws PistaDBException if the id is not found.
     */
    public synchronized void delete(long id) {
        checkOpen();
        nativeDelete(nativeHandle, id);
    }

    /**
     * Replaces the vector data for an existing id.
     *
     * @param vec Float array of length {@link #getDim()}.
     * @throws PistaDBException if the id is not found or dimensions mismatch.
     */
    public synchronized void update(long id, float[] vec) {
        checkOpen();
        nativeUpdate(nativeHandle, id, vec);
    }

    /**
     * Retrieves a stored vector and its label.
     *
     * @return {@link VectorEntry} containing the float array and label.
     * @throws PistaDBException if the id is not found.
     */
    public synchronized VectorEntry get(long id) {
        checkOpen();
        return nativeGet(nativeHandle, id);
    }

    /* ── Batch helpers ───────────────────────────────────────────────────── */

    /**
     * Inserts multiple vectors in one call.
     *
     * @param ids    Array of unique identifiers.
     * @param vecs   Array of float arrays, each of length {@link #getDim()}.
     * @param labels Optional array of labels (may be null; any element may be null).
     */
    public synchronized void insertBatch(long[] ids, float[][] vecs, String[] labels) {
        checkOpen();
        for (int i = 0; i < ids.length; i++) {
            String lbl = (labels != null && i < labels.length) ? labels[i] : "";
            nativeInsert(nativeHandle, ids[i], lbl != null ? lbl : "", vecs[i]);
        }
    }

    /** Inserts multiple vectors without labels. */
    public synchronized void insertBatch(long[] ids, float[][] vecs) {
        insertBatch(ids, vecs, null);
    }

    /* ── Search ──────────────────────────────────────────────────────────── */

    /**
     * K-nearest-neighbour search.
     *
     * @param query Float array of length {@link #getDim()}.
     * @param k     Number of results to return.
     * @return Array of up to {@code k} results ordered by ascending distance.
     * @throws PistaDBException on error (e.g. index not trained).
     */
    public synchronized SearchResult[] search(float[] query, int k) {
        checkOpen();
        return nativeSearch(nativeHandle, query, k);
    }

    /** Searches with k = 10. */
    public synchronized SearchResult[] search(float[] query) {
        return search(query, 10);
    }

    /* ── Index management ────────────────────────────────────────────────── */

    /**
     * Trains the index on currently inserted vectors.
     * <p>Required for {@link IndexType#IVF}, {@link IndexType#IVF_PQ},
     * and {@link IndexType#SCANN} before performing inserts.
     * Optional for {@link IndexType#HNSW} and {@link IndexType#DISKANN}
     * (triggers a rebuild pass).
     */
    public synchronized void train() {
        checkOpen();
        nativeTrain(nativeHandle);
    }

    /* ── Metadata ────────────────────────────────────────────────────────── */

    /** Returns the number of active (non-deleted) vectors. */
    public synchronized int getCount() {
        checkOpen();
        return nativeCount(nativeHandle);
    }

    /** Returns the vector dimensionality. */
    public synchronized int getDim() {
        checkOpen();
        return nativeDim(nativeHandle);
    }

    /** Returns the distance metric. */
    public synchronized Metric getMetric() {
        checkOpen();
        return Metric.fromValue(nativeMetric(nativeHandle));
    }

    /** Returns the index algorithm. */
    public synchronized IndexType getIndexType() {
        checkOpen();
        return IndexType.fromValue(nativeIndexType(nativeHandle));
    }

    /**
     * Returns the human-readable description of the last error.
     * Useful for debugging after a caught {@link PistaDBException}.
     */
    public synchronized String getLastError() {
        if (nativeHandle == 0) return "database is closed";
        return nativeLastError(nativeHandle);
    }

    /** Returns the PistaDB library version string (e.g. {@code "1.0.0"}). */
    public static String getVersion() {
        return nativeVersion();
    }

    /** Returns {@code true} if the database is open (handle not closed). */
    public boolean isOpen() {
        return nativeHandle != 0;
    }

    /* ── Internal ────────────────────────────────────────────────────────── */

    private void checkOpen() {
        if (nativeHandle == 0) {
            throw new PistaDBException("Database is closed");
        }
    }
}

File: ios/Sources/PistaDB/PistaDB.swift
/*
 *    ___ _    _        ___  ___
 *   | _ (_)__| |_ __ _|   \| _ )
 *   |  _/ (_-<  _/ _` | |) | _ \
 *   |_| |_/__/\__\__,_|___/|___/
 */
import Foundation
import PistaDBObjC

/// PistaDB – embedded vector database for iOS and macOS.
///
/// Wraps the Objective-C ``PSTDatabase`` layer which in turn calls the
/// underlying C library.  Use try-with `defer` or a ``withDatabase`` block
/// to ensure the handle is closed cleanly.
///
/// ## Quick-start
/// ```swift
/// let db = try PistaDB(path: dbPath, dim: 384, metric: .cosine)
/// defer { db.close() }
///
/// try db.insert(id: 1, vector: embedding, label: "My document")
/// let hits = try db.search(queryVec, k: 5)
/// try db.save()
/// ```
///
/// ## SwiftUI / async
/// ```swift
/// let results = try await db.search(queryVec, k: 10)
/// ```
public final class PistaDB {

    private let db: PSTDatabase

    // MARK: - Init

    /// Opens or creates a database with default parameters and HNSW index.
    ///
    /// - Parameters:
    ///   - path: Path to the `.pst` file (created if absent).
    ///   - dim:  Vector dimensionality.
    public convenience init(path: String, dim: Int) throws {
        try self.init(path: path, dim: dim, metric: .l2, indexType: .hnsw, params: nil)
    }

    /// Opens or creates a database.
    ///
    /// - Parameters:
    ///   - path:      Path to the `.pst` file (created if absent).
    ///   - dim:       Vector dimensionality.
    ///   - metric:    Distance metric (default `.l2`).
    ///   - indexType: Index algorithm (default `.hnsw`).
    ///   - params:    Tuning parameters; `nil` uses C-library defaults.
    public init(path: String,
                dim: Int,
                metric: Metric = .l2,
                indexType: IndexType = .hnsw,
                params: PistaDBParams? = nil) throws
    {
        var nsErr: NSError?
        guard let database = PSTDatabase(path: path,
                                         dim: dim,
                                         metric: metric.objc,
                                         indexType: indexType.objc,
                                         params: params?.toPST(),
                                         error: &nsErr)
        else {
            throw nsErr?.toPistaDBError() ?? PistaDBError.openFailed("unknown")
        }
        self.db = database
    }

    // MARK: - Lifecycle

    /// Persists the database to disk without closing the handle.
    public func save() throws {
        var nsErr: NSError?
        guard db.save(&nsErr) else { throw nsErr!.toPistaDBError() }
    }

    /// Saves then frees the native handle. Safe to call multiple times.
    public func close() { db.close() }

    // MARK: - Insert

    /// Inserts a vector with an optional label.
    ///
    /// - Parameters:
    ///   - id:     Unique identifier (user-managed).
    ///   - vector: Float array of length ``dim``.
    ///   - label:  Optional human-readable label (max 255 bytes).
    public func insert(id: UInt64, vector: [Float], label: String = "") throws {
        var nsErr: NSError?
        let ok = vector.withUnsafeBufferPointer {
            db.insert(id: id,
                      floatArray: $0.baseAddress!,
                      count: vector.count,
                      label: label,
                      error: &nsErr)
        }
        if !ok { throw nsErr!.toPistaDBError() }
    }

    // MARK: - Delete

    /// Soft-deletes a vector by id. Space is reclaimed on the next ``save()``.
    public func delete(id: UInt64) throws {
        var nsErr: NSError?
        guard db.delete(id: id, error: &nsErr) else { throw nsErr!.toPistaDBError() }
    }

    // MARK: - Update

    /// Replaces the vector data for an existing id.
    ///
    /// - Parameters:
    ///   - id:     Identifier to update.
    ///   - vector: Replacement float array of length ``dim``.
    public func update(id: UInt64, vector: [Float]) throws {
        var nsErr: NSError?
        let ok = vector.withUnsafeBufferPointer {
            db.update(id: id,
                      floatArray: $0.baseAddress!,
                      count: vector.count,
                      error: &nsErr)
        }
        if !ok { throw nsErr!.toPistaDBError() }
    }

    // MARK: - Get

    /// Retrieves a stored vector and its label.
    public func get(id: UInt64) throws -> VectorEntry {
        var nsErr: NSError?
        guard let entry = db.entry(id: id, error: &nsErr) else {
            throw nsErr!.toPistaDBError()
        }
        return VectorEntry(entry)
    }

    // MARK: - Batch insert

    /// Inserts multiple vectors in one call.
    ///
    /// - Parameters:
    ///   - entries: Sequence of `(id, vector, label)` triples.
    public func insertBatch<S: Sequence>(_ entries: S) throws
        where S.Element == (id: UInt64, vector: [Float], label: String)
    {
        for (id, vector, label) in entries {
            try insert(id: id, vector: vector, label: label)
        }
    }

    // MARK: - Search

    /// K-nearest-neighbour search.
    ///
    /// - Parameters:
    ///   - query: Float array of length ``dim``.
    ///   - k:     Number of results (default 10).
    /// - Returns: Up to `k` results ordered by ascending distance.
    public func search(_ query: [Float], k: Int = 10) throws -> [SearchResult] {
        var nsErr: NSError?
        let hits: NSArray? = query.withUnsafeBufferPointer {
            db.search(floatArray: $0.baseAddress!,
                      count: query.count,
                      k: k,
                      error: &nsErr) as NSArray?
        }
        guard let hits else { throw nsErr!.toPistaDBError() }
        return (hits as! [PSTSearchResult]).map(SearchResult.init)
    }

    // MARK: - Train

    /// Trains the index on currently inserted vectors.
    ///
    /// Required for `IVF`, `IVF_PQ`, and `ScaNN` before inserting.
    /// Optional for `HNSW` / `DiskANN` (triggers a rebuild pass).
    public func train() throws {
        var nsErr: NSError?
        guard db.train(&nsErr) else { throw nsErr!.toPistaDBError() }
    }

    // MARK: - Metadata

    /// Number of active (non-deleted) vectors.
    public var count: Int { db.count }

    /// Vector dimensionality.
    public var dim: Int { db.dim }

    /// Distance metric in use.
    public var metric: Metric { Metric(db.metric) }

    /// Index algorithm in use.
    public var indexType: IndexType { IndexType(db.indexType) }

    /// Human-readable description of the last error.
    public var lastError: String { db.lastError }

    /// Library version string, e.g. `"1.0.0"`.
    public static var version: String { PSTDatabase.version() }
}
